src/tpm2.rs

```diff
@@ -0,0 +1,957 @@
+const TPM_IO_ERROR: i32 = 5;
+const RETRY: usize = 4;
+
+static EMPTYMASK: &'static str = "1";
+
+#[derive(Debug, Clone)]
+struct TPM {
+    tpmdata: Value,
+    hash_alg: String,
+    encrypt_alg: String,
+    sign_alg: String,
+    legacy_tools: bool,
+}
+
+impl TPM {
+    pub fn new() -> TPM {
+        TPM {
+            tpmdata: json!(null),
+            hash_alg: String::from("sha256"),
+            encrypt_alg: String::from("rsa"),
+            sign_alg: String::from("rsassa"),
+            legacy_tools: tpm2::command_exist("tpm2_takeownership"),
+        }
+    }
+
+    fn read_tpmdata(&mut self) -> Result<(), KeylimeTpmError> {
+        self.tpmdata = File::open("tpmdata.json")
+            .and_then(|f| serde_json::from_reader(f).map_err(|e| e.into()))?;
+        Ok(())
+    }
+
+    fn write_tpmdata(&mut self) -> Result<(), KeylimeTpmError> {
+        let mut f = File::create("tpmdata.json")?;
+        let data_string: String =
+            serde_json::to_string_pretty(&self.tpmdata)?;
+        f.write_all(data_string.as_bytes())?;
+        Ok(())
+    }
+
+    pub fn update_tpmdata(
+        &mut self,
+        key: &str,
+        value: &str,
+    ) -> Result<(), KeylimeTpmError> {
+        match self.tpmdata.get_mut(key) {
+            Some(ptr) => *ptr = json!(value),
+            None => {
+                return Err(KeylimeTpmError::new_tpm_rust_error(
+                    format!("Key: {} is missing in tpmdata.json", key)
+                        .as_str(),
+                ));
+            }
+        };
+        Ok(())
+    }
+
+    // empty string means the value is not exist in the tpmdata.
+    pub fn get_tpmdata(&self, key: &str) -> String {
+        match self.tpmdata[key].as_str() {
+            None => String::new(),
+            Some(s) => s.to_string(),
+        }
+    }
+
+    pub fn init(
+        &mut self,
+        self_activate: bool,
+        config_pw: String,
+    ) -> Result<(String, String, String, String, String), KeylimeTpmError>
+    {
+        self.startup()?;
+        self.read_tpmdata()?;
+        self.emulator_warning();
+        self.set_password(config_pw)?;
+        self.create_ek(None)?;
+        self.get_pub_ek()?;
+        let ekcert = self.read_ekcert_nvram()?;
+
+        self.update_tpmdata("ekcert", &ekcert)?;
+        self.create_aik(self_activate)?;
+
+        let ek = self.get_tpmdata("ek");
+        let aik = self.get_tpmdata("aik");
+        let ek_tpm = self.get_tpmdata("ek_tpm");
+        let aik_name = self.get_tpmdata("aik_name");
+        Ok((ek, ekcert.into(), aik, ek_tpm, aik_name))
+    }
+
+    fn emulator_warning(&self) {
+        if self.is_software_tpm() {
+            warn!("{}{}{}{}",
+               "INSECURE: Keylime is using a software TPM emulator rather than",
+               "a real hardware TPM.\nINSECURE: The security of Keylime is NOT",
+               "linked to a hardware root of trust.\nINSECURE: Only use Keylim",
```

Comment:
Don't break words when wrapping text.  Check that spacing is preserved on other lines.

Reply:
You did not check that spacing is preserved when wrapping lines.